### PR TITLE
Make and use a julia system image

### PR DESCRIPTION
**Pull request by rwest**
This originally had in its scope some changes to how we use pyjulia, but that has been broken off into https://github.com/ReactionMechanismGenerator/RMG-Py/pull/2444

### Motivation or Problem
- Time to launch is awful (for me on my mac) because it recompiles a ton of julia stuff every time you import anything from reactors.py
- There's a way to launch Julia with a pre-compiled system image ([pyjulia docs](https://pyjulia.readthedocs.io/en/latest/troubleshooting.html#create-a-custom-system-image)) which should both get around the static linking problem inherent in conda's python, perhaps remove the need to use python-jl, and accelerate time to launch.
- Some of the code around this was a bit confusing.

### Description of Changes
- incorporates the changes from https://github.com/ReactionMechanismGenerator/RMG-Py/pull/2381  that advises developers to build a system image of julia with RMS compiled.
- refactors that code to be easier (for me) to understand

I am hoping that, before merging, this PR will get this working nicely and greatly accelerate time to launch.

### Testing
It is not yet ready to review for a merge, but I'm opening the pull request now (and requesting reviews)
- to start discussion
- for something for folks to hack on
- to see if it passes the CI tests

### Caveats
For me, on my intel Mac with julia and pyjulia coming from the `conda-forge` channel, it does **not** yet (2023-05-19) accelerate anything 😞 .  Oh, and the current instructions on building system image don't quite work.

### CODE REVIEW

**Review by JacksonBurns — CHANGES_REQUESTED** [review]
This will be a nice addition to RMG-Py once it is working - as of writing the conda-forge pyjulia causes build to fail because we have skipped the linking step present in our custom build:

`sed -i 's|bin/python|bin/python3|g' $(which python-jl)`

(from ChatGPT) This command uses the sed command to replace all occurrences of bin/python with bin/python3 in the file specified by $(which python-jl). The -i option modifies the file in place. So, when you run this command, it will replace all instances of bin/python with bin/python3 in the specified file.

I think we can just run this command after solving the conda environment and get the same effect? Or would building the system image remove the need to do so?

I have also left some stylistic comments.

The CI should build a Julia image so that this use case is actually tested, yes?

**Comment by JacksonBurns** [comment]
Possibly unhelpful idea: [`PythonCall.jl`](https://github.com/cjdoris/PythonCall.jl) is an alternative to PyJulia that is better maintained. They have a page to help [move to their tool](https://cjdoris.github.io/PythonCall.jl/stable/pycall/).
Can we consider doing so?
Would doing so require lots of editing RMS? And potentially also pyrms? Or even RMG's `reactors.py`?
Why do they call it rush hour when nothing moves?

File: rmgpy/rmg/reactors.py [hunk 1]
```diff
@@ -35,25 +35,33 @@
...
-    try:
-        from pyrms import rms
-        from diffeqpy import de
-        from julia import Main
-    except:
+from os.path import dirname, abspath, join, exists
+
+try:
+    from julia.api import Julia
+    system_image_path = join(dirname(dirname(dirname(abspath(__file__)))),"rms.so")
```
[the remark this answers is not in the document]
Comment (rwest) [reply rmgpy/rmg/reactors.py]: I use rmgpy.get_path in  https://github.com/ReactionMechanismGenerator/RMG-Py/pull/2443/commits/841543e867727695eaf3bc5cb98ff0e13fb6f705

Strangely, it's not (for me, yet) much faster. If/when it works well it should absolutely be default.

File: rmgpy/rmg/reactors.py [hunk 2]
```diff
@@ -35,25 +35,33 @@
...
+    if exists(system_image_path):
+        jl = Julia(sysimage=system_image_path)
+    elif __debug__:
+        "This is the normal case (__debug__ is True by default)"
```
[the remark this answers is not in the document]
Comment (rwest) [reply rmgpy/rmg/reactors.py]: Done in https://github.com/ReactionMechanismGenerator/RMG-Py/pull/2443/commits/5169d108982e378237940a49fe327bfaf536a12b

File: rmgpy/rmg/reactors.py
[the remark this answers is not in the document]
Comment (rwest) [reply rmgpy/rmg/reactors.py:1]: Raising an exception to be handled by the calling code would mean the rest of this file would not get imported, which is what this "try/pass" thing is trying to achieve.  I think the use case is code that needs to import this module, but doesn't actually execute anything important in it.
Not very satisfactory, I agree, (in more ways than one) but fixing is beyond scope of this PR.

File: documentation/source/users/rmg/installation/anacondaDeveloper.rst [hunk 3]
```diff
@@ -188,3 +264,15 @@ A number of basic examples can be run immediately.  Additional example input fil
 
     cd RMG-Py
     make eg4
+
+
+Building Documentation
+======================
+To build the documentation (to test that you have it right before pushing to Github) you will need to install sphinx::
```
[the remark this answers is not in the document]
Comment (rwest) [reply documentation/source/users/rmg/installation/anacondaDeveloper.rst]: Fixed in https://github.com/ReactionMechanismGenerator/RMG-Py/pull/2443/commits/a358bfa750da9990d9145e8635d029a2bdb33fca

File: rmgpy/rmg/reactors.py [hunk 4]
```diff
@@ -35,25 +35,33 @@
...
+        I don't know why or how often that is done (all it does is
+        remove assert statements), nor why one would then
+        need or wish to skip creating a Julia runtime, but for now I'm trying
+        to refactor without changing behaviour.
+        """
         pass
+    from pyrms import rms
+    from diffeqpy import de
+    from julia import Main
+except Exception as e:
```
[the remark this answers is not in the document]
Comment (rwest) [reply rmgpy/rmg/reactors.py:65]: I agree in principle, bare exceptions are a code smell. But I'm hoping that in future this whole thing can go away.  I think it was introduced so ARC could pass nose tests or something - where this file is imported somewhere but nothing in it needs to actually run.
For this use case, all exceptions should be treated the same, so I'm going to leave it as is, until we get rid of it entirely.
At least I added some logging so we can see what the exception is.

**Pull request closed without merging**